### PR TITLE
v8 Pixel Rounding

### PR DESCRIPTION
**Pull request by GoodBoyDigital**
This PR bring roundPixels to v8.  It now does this directly into the shader. This is a new approach made possible by the fact we have a shader compiler!

The main advantages of this approach: 

- less work on the cpu rounding all those verts!
- MUCH more accurate pixel snapping - no matter the transform applied, as it all happes right before the position is sent to the frag shader.
- no change to upload size as round pixels prop is packed into the textureID variable. As both numbers are low, they can be packed in as 2 uint16s.

v7 rounding:
image

v8 rounding:
image

### Summary
🖌️🔬🚀

This pull request adds a new feature to the rendering system that allows users to enable or disable pixel rounding for sprites and graphics. Pixel rounding can improve the visual quality and performance of rendering by avoiding sub-pixel rendering and texture filtering. The feature is implemented by adding a `roundPixels` option to the `AbstractRenderer` class and a `roundPixels` property to the `View` interface. The feature is also supported by the high-shader module, which uses a new `roundPixelsBit` shader bit and a `uResolution` global uniform. The feature is integrated with the batcher module, which uses a new `BatchGeometry` class and a `aTextureIdAndRound` vertex attribute. The feature is also compatible with the mask and layer modules, which use a new `projectionData` field instead of `projectionMatrix`. The pull request also fixes some issues and cleans up some code in the rendering system.

> _Sing, O Muse, of the glorious deeds of the coder_
> _Who with skill and wisdom enhanced the renderer_
> _And made the sprites and graphics shine with round pixels_
> _Like the stars that adorn the night sky of Olympus_

### Walkthrough
*  Add round pixels logic to the shaders for WebGPU and WebGL rendering ([link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-71719a370a41477b6b10f352b538614bc2982d420ca5c2c2bdd23365aeed69ccR6), [link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-71719a370a41477b6b10f352b538614bc2982d420ca5c2c2bdd23365aeed69ccR44), [link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-a0f3f9b4b6542d13dbae01459e1936d490c982d65be81867d1257e05996b0a19R5), [link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-a0f3f9b4b6542d13dbae01459e1936d490c982d65be81867d1257e05996b0a19R39), [link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-2cace2a5814aafcb78a26a7a1cbf4fba8816d64d9d06ddded973d0581969df73R39-R40), [link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-2cace2a5814aafcb78a26a7a1cbf4fba8816d64d9d06ddded973d0581969df73L96-R98), [link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-7daa937165387d457eda861ee4c32d5f12f77b321927bd618b34d591ed5af9c9R5-R9), [link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-7daa937165387d457eda861ee4c32d5f12f77b321927bd618b34d591ed5af9c9L68-R83), [link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-7daa937165387d457eda861ee4c32d5f12f77b321927bd618b34d591ed5af9c9R105-R109), [link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-7daa937165387d457eda861ee4c32d5f12f77b321927bd618b34d591ed5af9c9L126-R153), [link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-79f1b1ad1ca89b3da9175d612bf07508450aa908fc09342517b804b137c67211L8-R9), [link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-79f1b1ad1ca89b3da9175d612bf07508450aa908fc09342517b804b137c67211R25), [link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-04d4339f0c35c484d60d636c4880553bdb246c7e4b61806ed6c058fdb1329f0dR9), [link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-04d4339f0c35c484d60d636c4880553bdb246c7e4b61806ed6c058fdb1329f0dR17-R22), [link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-04d4339f0c35c484d60d636c4880553bdb246c7e4b61806ed6c058fdb1329f0dL27-R44), [link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-8e88e5072a2c68f3e6bc7e494e5124658d831651c7d2409611bdf4b3cf98ee0fR1-R24))
*  Modify the BatchGeometry class to store the round pixels flag in the aTextureIdAndRound attribute for each vertex ([link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-ef441d7af72c6d0d04e6dfddfb4b0f55333f391a84bd54ad28386728bfb7955dL51-R54), [link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-7daa937165387d457eda861ee4c32d5f12f77b321927bd618b34d591ed5af9c9L68-R83), [link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-7daa937165387d457eda861ee4c32d5f12f77b321927bd618b34d591ed5af9c9L126-R153))
*  Modify the Batcher class to store the round pixels flag for each batch ([link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-5b0990b7f67a73bcf39db49902b4f21c89d78fdb3a0077293ddcf822eb047341R78-R79), [link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-5b0990b7f67a73bcf39db49902b4f21c89d78fdb3a0077293ddcf822eb047341L101))
*  Modify the BatcherPipe class to use the BatchGeometry class instead of the Geometry class for batching sprites and graphics ([link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-89625f6f630aefb661146d12dc5859848581a3df90cbcf736ec1094c4c97d3a8L3-R3), [link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-89625f6f630aefb661146d12dc5859848581a3df90cbcf736ec1094c4c97d3a8L37-R37), [link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-89625f6f630aefb661146d12dc5859848581a3df90cbcf736ec1094c4c97d3a8L58-R58))
*  Modify the GlobalUniformSystem class to store and bind the resolution of the render target as a global uniform for the shaders ([link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-df7c6756a9c195bfecfc594cc0358bbf6dcf864aaea4667ef2942e9ff4efb278L20-R25), [link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-df7c6756a9c195bfecfc594cc0358bbf6dcf864aaea4667ef2942e9ff4efb278R36), [link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-df7c6756a9c195bfecfc594cc0358bbf6dcf864aaea4667ef2942e9ff4efb278L102-R104), [link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-df7c6756a9c195bfecfc594cc0358bbf6dcf864aaea4667ef2942e9ff4efb278L108-R113), [link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-df7c6756a9c195bfecfc594cc0358bbf6dcf864aaea4667ef2942e9ff4efb278L116-R121), [link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-df7c6756a9c195bfecfc594cc0358bbf6dcf864aaea4667ef2942e9ff4efb278R135-R137), [link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-df7c6756a9c195bfecfc594cc0358bbf6dcf864aaea4667ef2942e9ff4efb278R197))
*  Modify the RenderTarget class to store and expose the size of the render target in pixels ([link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-800463114c8f3b95a93ab177292457fb7733250e0247b3c2f3c043e937383fa6R50), [link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-800463114c8f3b95a93ab177292457fb7733250e0247b3c2f3c043e937383fa6R105-R114))
*  Modify the AbstractRenderer class to accept and expose the roundPixels option for the renderer ([link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-27fa29afe053bf0bbdb6c29016c4da0abb00506b5b31c3e689aa84f35867aa42R66-R68), [link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-27fa29afe053bf0bbdb6c29016c4da0abb00506b5b31c3e689aa84f35867aa42L95-R98), [link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-27fa29afe053bf0bbdb6c29016c4da0abb00506b5b31c3e689aa84f35867aa42R108-R109), [link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-27fa29afe053bf0bbdb6c29016c4da0abb00506b5b31c3e689aa84f35867aa42R356-R360))
*  Modify the View interface to include the roundPixels property for the view ([link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-29df9db790c307f0ec1a316f1382f0addcab8b48ad5e921871f86d6f097d7748R28-R33))
*  Modify the GlGraphicsAdaptor class to use the roundPixelsBitGl value and expose the shader property ([link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-c68056a5ead5ba5ee16889ab55d50dd7805b64f77e4956a31cb9143163da7108L8-R11), [link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-c68056a5ead5ba5ee16889ab55d50dd7805b64f77e4956a31cb9143163da7108L29-R29), [link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-c68056a5ead5ba5ee16889ab55d50dd7805b64f77e4956a31cb9143163da7108L44-R48), [link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-c68056a5ead5ba5ee16889ab55d50dd7805b64f77e4956a31cb9143163da7108L59-R67), [link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-c68056a5ead5ba5ee16889ab55d50dd7805b64f77e4956a31cb9143163da7108L115-R98))
*  Modify the GpuGraphicsAdaptor class to use the roundPixelsBit value and expose the shader property ([link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-6d61decaf9f7217bf04dfe80ada52a992df26f97bfd786019c1afdc29ceaa41cL8-R10), [link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-6d61decaf9f7217bf04dfe80ada52a992df26f97bfd786019c1afdc29ceaa41cL31-R64), [link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-6d61decaf9f7217bf04dfe80ada52a992df26f97bfd786019c1afdc29ceaa41cL45-R83), [link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-6d61decaf9f7217bf04dfe80ada52a992df26f97bfd786019c1afdc29ceaa41cL61-R96), [link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-6d61decaf9f7217bf04dfe80ada52a992df26f97bfd786019c1afdc29ceaa41cL103-R115), [link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-6d61decaf9f7217bf04dfe80ada52a992df26f97bfd786019c1afdc29ceaa41cL133-R146))
*  Modify the AlphaMaskPipe, LayerPipe, and LayerSystem classes to use the projectionData object instead of the projectionMatrix field ([link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-fa0a43dfb73e1c4b162870aae70ee63af210695afa46b14e1e947f96eed034b1L171-R171), [link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-c3f5ac7aff40f9c9fac60e3321301d395d6c1652533fd7353b5732ac4ef46472L39-R39), [link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-0d16cb5973b93e4603e2f89fbf8467e78da6962d56823d8115eba80d72e8e95dL93-R93))
*  Remove unused imports and comments from the Batcher and BatcherPipe classes ([link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-5b0990b7f67a73bcf39db49902b4f21c89d78fdb3a0077293ddcf822eb047341L4), [link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-5b0990b7f67a73bcf39db49902b4f21c89d78fdb3a0077293ddcf822eb047341L13-L14), [link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-89625f6f630aefb661146d12dc5859848581a3df90cbcf736ec1094c4c97d3a8L3-R3))
*  Delete the `src/rendering/batcher/gpu/getBatchedGeometry.ts` file as it is no longer used ([link](https://github.com/pixijs/pixijs/pull/9714/files?diff=unified&w=0#diff-a5da1e203f83efc2d340448ac265bad3e66309f8b91c1e670b3e1741847ccc97))

### CODE REVIEW

**Comment by bigtimebuddy** [comment]
I'm not sure the intention with roundPixels is to also give Text an aliased look, it's only to avoid rendering elements on sub-pixels. I think the Text thing is a cool feature, but is something different.

File: src/utils/data/uid.ts [hunk 1]
```diff
@@ -7,23 +7,23 @@ type UIDNames =
...
-    | 'graphicsPath'
-    | 'fillGradient'
-    | 'fillPattern'
-    | 'meshView'
-    | 'renderable'
-    | 'buffer'
-    | 'bufferResource'
+    | 'batcher' //
+    | 'graphicsContext' //
+    | 'graphicsView' //
```
Comment (Zyie) [inline src/utils/data/uid.ts:12]: What is this for?

Comment (GoodBoyDigital) [reply src/utils/data/uid.ts:12]: any // we can remove - but did not want to muddy the PR!

**Comment by GoodBoyDigital** [comment]
> I'm not sure the intention with roundPixels is to also give Text an aliased look, it's only to avoid rendering elements on sub-pixels. I think the Text thing is a cool feature, but is something different.

no worries - we are on the same page. The aliased look is because the image is scaled up - so its just showing the canvas pixel as perfectly aligned

File: tests/renderering/RoundPixels.test.ts [hunk 2]
```diff
@@ -15,23 +16,13 @@ describe('Round Pixels', () =>
 {
     it('Round pixels should set correctly when passed as an option', async () =>
```
Comment (GoodBoyDigital) [inline tests/renderering/RoundPixels.test.ts:17]: i think this is a different test. Perhaps we have both?

one is testing the renderer and one is testing the Application

File: src/rendering/renderers/shared/system/AbstractRenderer.ts [hunk 3]
```diff
@@ -353,9 +353,4 @@ export class AbstractRenderer<PIPES, OPTIONS>
     {
         return this.textureGenerator.generateTexture(options);
     }
-
-    get roundPixels(): boolean
```
Comment (GoodBoyDigital) [inline src/rendering/renderers/shared/system/AbstractRenderer.ts]: use might want to log this?

File: src/rendering/high-shader/shader-bits/roundPixelsBit.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    }
+};
+
+export const roundPixelsBitGl = {
+    name: 'round-pixels-bit',
+    vertex: {
+        header: /* glsl */`   
+            vec2 roundPixels(vec2 position, vec2 targetSize)
+            {       
+                return (round((position * 0.5 + 0.5) * targetSize) / targetSize) * 2.0 - 1.0;
```
Comment (GoodBoyDigital) [inline src/rendering/high-shader/shader-bits/roundPixelsBit.ts:19]:
```suggestion
                return (floor((position * 0.5 + 0.5) * targetSize) / targetSize) * 2.0 - 1.0;
```

**Review by Zyie — APPROVED** [review]

**Pull request merged**